ush/forecast_postdet.sh

```diff
@@ -1038,7 +1038,7 @@ GOCART_postdet() {
       rm -f "${COM_CHEM_HISTORY}/gocart.inst_aod.${vdate:0:8}_${vdate:8:2}00z.nc4"
     fi
 
-    ${NLN} "${COM_CHEM_HISTORY}/gocart.inst_aod.${vdate:0:8}_${vdate:8:2}00z.nc4" \
-           "${DATA}/gocart.inst_aod.${vdate:0:8}_${vdate:8:2}00z.nc4"
+    #${NLN} "${COM_CHEM_HISTORY}/gocart.inst_aod.${vdate:0:8}_${vdate:8:2}00z.nc4" \
```

Comment:
@WalterKolczynski-NOAA @bbakernoaa @lipan-NOAA @rmontuoro  --- I think i have other things working in my other branch so I'm coming back to other issues including this.  Is it anticipated that if I link these files that the model wouldn't run?  Barry helped me figure out that this was causing some earlier crashes.  I can re-test to see if this is still an issue, but want to see if there's some known issue with this or suggested workaround.  My other branch is : https://github.com/JessicaMeixner-NOAA/global-workflow/tree/feature/updateufsstack15

Reply:
Without the links, the output will never get into a permanent location. I think what needs to happen is any existing files at the target need to be deleted. GOCART seems to be okay with the links as long as the target doesn't exist, otherwise we would be seeing more problems.

Reply:
Okay I can try again without this commented out and see if other issuses I was having was partially disguised as this

Reply:
Uncommenting  this line resulted in a failure still (even from a completely clean run): 
```
nid001088.dogwood.wcoss2.ncep.noaa.gov 0: UFS Aerosols: Advancing from 2021-03-23T17:40:00 to 2021-03-23T18:00:00
nid001088.dogwood.wcoss2.ncep.noaa.gov 0:
nid001088.dogwood.wcoss2.ncep.noaa.gov 0:  Writing:     28 Slices to File:  gocart.inst_aod.20210323_1800z.nc4
nid001088.dogwood.wcoss2.ncep.noaa.gov 0: pe=00000 FAIL at line=00187    NetCDF4_FileFormatter.F90                <status=13>
pe=00000 FAIL at line=00061    HistoryCollection.F90                    <status=13>
pe=00000 FAIL at line=00790    ServerThread.F90                         <status=13>
pe=00000 FAIL at line=00138    BaseServer.F90                           <status=13>
pe=00000 FAIL at line=00981    ServerThread.F90                         <status=13>
pe=00000 FAIL at line=00094    MessageVisitor.F90                       <status=13>
pe=00000 FAIL at line=00113    AbstractMessage.F90                      <status=13>
pe=00000 FAIL at line=00107    SimpleSocket.F90                         <status=13>
pe=00000 FAIL at line=00429    ClientThread.F90                         <status=13>
pe=00000 FAIL at line=00363    ClientManager.F90                        <status=13>
pe=00000 FAIL at line=03524    MAPL_HistoryGridComp.F90                 <status=13>
pe=00000 FAIL at line=01818    MAPL_Generic.F90                         <status=13>
pe=00000 FAIL at line=01284    MAPL_CapGridComp.F90                     <status=13>
pe=00000 FAIL at line=01213    MAPL_CapGridComp.F90                     <status=13>
pe=00000 FAIL at line=01159    MAPL_CapGridComp.F90                     <status=13>
pe=00000 FAIL at line=00827    MAPL_CapGridComp.F90                     <status=13>
pe=00000 FAIL at line=00967    MAPL_CapGridComp.F90                     <status=13>
nid001088.dogwood.wcoss2.ncep.noaa.gov 0: MPICH ERROR [Rank 0] [job id 2794f105-9f31-4db4-b1dc-56a1883195f6] [Fri Oct 13 12:32:25 2023] [nid001088] - Abort(1) (rank 0 in comm 496): application called MPI_Abort(comm=0x84000002, 1) - process 0

nid001088.dogwood.wcoss2.ncep.noaa.gov 0: forrtl: severe (174): SIGSEGV, segmentation fault occurred
Image              PC                Routine            Line        Source
ufs_model.x        000000000641185A  Unknown               Unknown  Unknown
libpthread-2.31.s  00001456DAA868C0  Unknown               Unknown  Unknown
libmpi_intel.so.1  00001456DCBE3BFA  Unknown               Unknown  Unknown
libmpi_intel.so.1  00001456DCABC05F  Unknown               Unknown  Unknown
libmpi_intel.so.1  00001456DB1C9DA8  MPI_Abort             Unknown  Unknown
ufs_model.x        00000000013E79C4  _ZN5ESMCI3VMK5abo         757  ESMCI_VMKernel.C
ufs_model.x        00000000013C6B57  _ZN5ESMCI2VM5abor        3597  ESMCI_VM.C
ufs_model.x        0000000000BE8E83  c_esmc_vmabort_          1190  ESMCI_VM_F.C
ufs_model.x        000000000054D279  esmf_vmmod_mp_esm        9431  ESMF_VM.F90
ufs_model.x        00000000006CEFCF  esmf_initmod_mp_e        1226  ESMF_Init.F90
ufs_model.x        000000000042B7B0  MAIN__                    403  UFS.F90
ufs_model.x        000000000042A292  Unknown               Unknown  Unknown
libc-2.31.so       00001456DA69124D  __libc_start_main     Unknown  Unknown
```
Full log file: /lfs/h2/emc/couple/noscrub/jessica.meupdatemodel/s2swc48t03/COMROOT/s2swc48t03/logs/2021032312/gfsfcst.log.0


It does seem that GOCART has a problem with this, unless I'm missing something.  At this point I'm ready to do a fresh round of low res testing + a high res spot check and open a new PR to update the model.   But this seems to likely be a sticking point.    I'm hoping someone who works on the aerosols component can chime in on this.

Reply:
For now, comment it out and add a `GOCART_out()` to match the others that copies the files to `COM_CHEM_HISTORY` at the end of the forecast.

I'd also like to know what changed that this no longer works and if there is anyone working to change it back.

Reply:
Will do.   

Hopefully someone who works on the aerosol component can chime on your second question.

Reply:
@WalterKolczynski-NOAA multiple attempts to add this to the copy at the end have failed.  

Current attempt log: 
/scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/COMROOT/s2swc48t03/logs/2021032312/gfsfcst.log.0

error: 
```
SUB GOCART_out: Copying output data for GOCART
+ forecast_postdet.sh[1052]: /bin/cp -p '/scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/RUNDIRS/s2swc48t03/fcst.244373/gocart.inst_aod.*' /scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/COMROOT/s2swc48t03/gfs.20210323/12//model_data/chem/history/
/bin/cp: cannot stat '/scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/RUNDIRS/s2swc48t03/fcst.244373/gocart.inst_aod.*': No such file or directory
```

which is odd since these files do indeed exist: 
```
ls /scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/RUNDIRS/s2swc48t03/fcst.244373/gocart.inst_aod.*
/scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/RUNDIRS/s2swc48t03/fcst.244373/gocart.inst_aod.20210323_1800z.nc4
/scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/RUNDIRS/s2swc48t03/fcst.244373/gocart.inst_aod.20210324_0000z.nc4
/scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/RUNDIRS/s2swc48t03/fcst.244373/gocart.inst_aod.20210324_0600z.nc4
/scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/RUNDIRS/s2swc48t03/fcst.244373/gocart.inst_aod.20210324_1200z.nc4
/scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/RUNDIRS/s2swc48t03/fcst.244373/gocart.inst_aod.20210324_1800z.nc4
/scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/RUNDIRS/s2swc48t03/fcst.244373/gocart.inst_aod.20210325_0000z.nc4
/scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/RUNDIRS/s2swc48t03/fcst.244373/gocart.inst_aod.20210325_0600z.nc4
/scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/RUNDIRS/s2swc48t03/fcst.244373/gocart.inst_aod.20210325_1200z.nc4
/scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/RUNDIRS/s2swc48t03/fcst.244373/gocart.inst_aod.20210325_1800z.nc4
/scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/RUNDIRS/s2swc48t03/fcst.244373/gocart.inst_aod.20210326_0000z.nc4
/scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/RUNDIRS/s2swc48t03/fcst.244373/gocart.inst_aod.20210326_0600z.nc4
/scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/RUNDIRS/s2swc48t03/fcst.244373/gocart.inst_aod.20210326_1200z.nc4
/scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/RUNDIRS/s2swc48t03/fcst.244373/gocart.inst_aod.20210326_1800z.nc4
/scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/RUNDIRS/s2swc48t03/fcst.244373/gocart.inst_aod.20210327_0000z.nc4
/scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/RUNDIRS/s2swc48t03/fcst.244373/gocart.inst_aod.20210327_0600z.nc4
/scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/RUNDIRS/s2swc48t03/fcst.244373/gocart.inst_aod.20210327_1200z.nc4
/scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/RUNDIRS/s2swc48t03/fcst.244373/gocart.inst_aod.20210327_1800z.nc4
/scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/RUNDIRS/s2swc48t03/fcst.244373/gocart.inst_aod.20210328_0000z.nc4
/scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/RUNDIRS/s2swc48t03/fcst.244373/gocart.inst_aod.20210328_0600z.nc4
/scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t03/RUNDIRS/s2swc48t03/fcst.244373/gocart.inst_aod.20210328_1200z.nc4
```

Global-workflow clone:   /scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/global-workflow

github: https://github.com/JessicaMeixner-NOAA/global-workflow/tree/updateUFS101223

Reply:
You can't put the glob (`*`) inside parenthesis, as it won't expand. I think we would rather copy an explicit list instead anyway; you can just copy the loop from the old section.

Reply:
I understand wanting to copy explicit lists but we should really be general as possible here as the inst_aod is just one of the output files that are possible.

It would really be better if we copied or linked all of the `gocart.*.nc4` files to the `chem` directory as there are lots of possible diagnostics available

Reply:
Let me see if I can dig-out my error statement with trying to do the old block but make it a copy and reverse order... 

Currently trying this: 

+  if (( $( ls -1 "${DATA}/gocart.inst_aod."* 2> /dev/null | wc -l) > 0 )); then
+    for file in $(ls "${DATA}/gocart.inst_aod."*) ; do
+      ${NCP} "${file}"  "${COM_CHEM_HISTORY}/"
+    done
+  fi
+
+

Forecast is running to see if this works.

Reply:
The error from: /scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t02/COMROOT/s2swc48t02/logs/2021032312/gfsfcst.log.0
is: 

```
SUB GOCART_out: Copying output data for GOCART
+ forecast_postdet.sh[1052]: for fhr in '${FV3_OUTPUT_FH}'
++ forecast_postdet.sh[1053]: date --utc -d '20210323 12 + 0 hours' +%Y%m%d%H
+ forecast_postdet.sh[1053]: local vdate=2021032312
+ forecast_postdet.sh[1054]: /bin/cp -p /scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t02/RUNDIRS/s2swc48t02/fcst.123448/gocart.inst_aod.20210323_1200z.nc4 /scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t02/COMROOT/s2swc48t02/gfs.20210323/12//model_data/chem/history/gocart.inst_aod.20210323_1200z.nc4
/bin/cp: cannot stat '/scratch1/NCEPDEV/climate/Jessica.Meixner/HR3/updatemodel02/s2swc48t02/RUNDIRS/s2swc48t02/fcst.123448/gocart.inst_aod.20210323_1200z.nc4': No such file or directory
```
This was the code: https://github.com/JessicaMeixner-NOAA/global-workflow/commit/6a61d8a213ff07176358381b9c6e3a910a72b25f

Reply:
> I understand wanting to copy explicit lists but we should really be general as possible here as the inst_aod is just one of the output files that are possible.
> 
> It would really be better if we copied or linked all of the `gocart.*.nc4` files to the `chem` directory as there are lots of possible diagnostics available


We can't link them or the run dies.  If what I'm trying now works, we could try to make it slightly more general as long as it doesn't conflict with other linking statements.

Reply:
I think you need to skip the f000 one.

Reply:
There might be something in there that we just need to add to the [AERO_HISTORY.rc](https://github.com/JessicaMeixner-NOAA/global-workflow/blob/feature/updateufs/parm/ufs/chem/AERO_HISTORY.rc#L9) file.

It needs to be added at the top of the file 

`Allow_Overwrite: true`

Testing now

Reply:
How'd your tests go @bbakernoaa ?  Mine did not go well.  The changes I tried are here: https://github.com/JessicaMeixner-NOAA/global-workflow/tree/trygocartfix with the allow overwrite and back tracking the other changes. 

I also haven't had good luck copying files at the end of the run, I keep getting errors, that branch is here: https://github.com/JessicaMeixner-NOAA/global-workflow/tree/updateUFS101223